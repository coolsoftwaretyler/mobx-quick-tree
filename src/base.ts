import type { IAnyType as MSTAnyType } from "mobx-state-tree";
import { $quickType } from "./symbols";
import type { IAnyStateTreeNode, IStateTreeNode, TreeContext, StateTreeNode } from "./types";

export abstract class BaseType<InputType, OutputType, InstanceType> {
  /** @internal */
  readonly [$quickType] = undefined;

  readonly InputType!: InputType;
  readonly OutputType!: OutputType;
  readonly InstanceType!: StateTreeNode<InstanceType, this>;
  readonly InstanceTypeWithoutSTN!: InstanceType;

  readonly name: string;
  readonly mstType!: MSTAnyType;

  constructor(mstType: MSTAnyType) {
    this.name = mstType.name;

    Object.defineProperty(this, "mstType", {
      value: mstType,
      enumerable: false,
      writable: false,
      configurable: false,
    });
  }

  /**
   * Create a new instance of this class model in observable mode. Uses an `mobx-state-tree` type under the hood.
   */
  create(snapshot?: this["InputType"], env?: any): this["InstanceType"] {
    return this.mstType.create(snapshot, env);
  }

  abstract is(value: IAnyStateTreeNode): value is this["InstanceType"];
  abstract is(value: any): value is this["InputType"] | this["InstanceType"];

  /**
   * Create a new instance of this class model in readonly mode. Properties and views are accessible on readonly instances but actions will throw if run.
   */
  createReadOnly(snapshot?: InputType, env?: any): this["InstanceType"] {
    const context: TreeContext = {
      referenceCache: new Map(),
      referencesToResolve: [],
      env,
    };

    const instance = this.instantiate(snapshot, context, null);
    for (const resolver of context.referencesToResolve) {
      resolver();
    }

    return instance;
  }

  abstract instantiate(snapshot: this["InputType"] | undefined, context: TreeContext, parent: IStateTreeNode | null): this["InstanceType"];

  abstract schemaHash(): Promise<string>;
}
