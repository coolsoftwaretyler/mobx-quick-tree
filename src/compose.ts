import { types as mstTypes } from "mobx-state-tree";
import type { ModelInitializer } from "./model";
import { ModelType } from "./model";
import type { IAnyNodeModelType, INodeModelType } from "./types";

type PropsFromTypes<T> =
  T extends INodeModelType<infer P, any> ? P : T extends [INodeModelType<infer P, any>, ...infer Tail] ? P & PropsFromTypes<Tail> : {};

type OthersFromTypes<T> =
  T extends INodeModelType<any, infer O> ? O : T extends [INodeModelType<any, infer O>, ...infer Tail] ? O & OthersFromTypes<Tail> : {};

type ComposeFactory = {
  <Types extends [IAnyNodeModelType, ...IAnyNodeModelType[]]>(
    name: string,
    ...types: Types
  ): INodeModelType<PropsFromTypes<Types>, OthersFromTypes<Types>>;
  <Types extends [IAnyNodeModelType, ...IAnyNodeModelType[]]>(
    ...types: Types
  ): INodeModelType<PropsFromTypes<Types>, OthersFromTypes<Types>>;
};

export const compose: ComposeFactory = (nameOrType: IAnyNodeModelType | string, ...types: IAnyNodeModelType[]): IAnyNodeModelType => {
  let name: string | undefined = undefined;
  if (typeof nameOrType == "string") {
    name = nameOrType;
  } else {
    types.unshift(nameOrType);
    name = nameOrType.name;
  }

  const props = types.reduce((props, model) => ({ ...props, ...model.properties }), {});
  const initializers = types.reduce<ModelInitializer[]>((inits, model) => (model as ModelType<any, any>).initializers.concat(inits), []);

  // We ignore the overloading MST has put on compose, to avoid writing out an annoying `switch`
  const mstComposedModel = (mstTypes.compose as any)(name, ...types.map((t) => t.mstType));

  // TODO see if there's a good way to not have to do this cast
  return new ModelType(props, initializers, mstComposedModel);
};
