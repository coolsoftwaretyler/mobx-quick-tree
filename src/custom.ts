import type { CustomTypeOptions } from "mobx-state-tree";
import { types } from "mobx-state-tree";
import { BaseType } from "./base";
import type { IStateTreeNode, IType, TreeContext } from "./types";

export class CustomType<InputType, OutputType> extends BaseType<InputType, OutputType, OutputType> {
  constructor(readonly options: CustomTypeOptions<InputType, OutputType>) {
    super(types.custom<InputType, OutputType>(options));
  }

  instantiate(snapshot: InputType, context: TreeContext, parent: IStateTreeNode | null): this["InstanceType"] {
    if (snapshot === undefined) {
      throw new Error("can't initialize custom type with undefined");
    }
    return this.options.fromSnapshot(snapshot) as this["InstanceType"];
  }

  is(value: any): value is this["InstanceType"];
  is(value: any): value is this["InputType"] | this["InstanceType"] {
    return this.mstType.is(value);
  }

  async schemaHash() {
    return `custom:${this.options.name}`;
  }
}

export const custom = <InputType, OutputType>(
  options: CustomTypeOptions<InputType, OutputType>,
): IType<InputType, OutputType, OutputType> => {
  return new CustomType(options);
};
