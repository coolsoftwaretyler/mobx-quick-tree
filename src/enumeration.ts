import { types } from "mobx-state-tree";
import { BaseType } from "./base";
import type { IAnyStateTreeNode, TreeContext, ISimpleType, IStateTreeNode } from "./types";
import memoize from "lodash.memoize";

class EnumerationType<EnumOptions extends string> extends BaseType<EnumOptions, EnumOptions, EnumOptions> {
  constructor(
    readonly name: string,
    readonly options: readonly EnumOptions[],
  ) {
    super(types.enumeration<EnumOptions>(name, [...options]));
  }

  instantiate(snapshot: this["InputType"], _context: TreeContext, _parent: IStateTreeNode | null): this["InstanceType"] {
    if (typeof snapshot == "string" && this.options.includes(snapshot)) {
      return snapshot as this["InstanceType"];
    }
    throw new Error(`Unknown enum value \`${snapshot}\`. Options are: ${this.options.join(", ")}`);
  }

  is(value: IAnyStateTreeNode): value is this["InstanceType"];
  is(value: any): value is this["InputType"] | this["InstanceType"] {
    return this.options.includes(value);
  }

  schemaHash: () => Promise<string> = memoize(async () => {
    return `enum:${this.options.join("|")}`;
  });
}

type EnumerationFactory = {
  <EnumOptions extends string>(name: string, options: readonly EnumOptions[]): ISimpleType<EnumOptions>;
  <EnumOptions extends string>(options: readonly EnumOptions[]): ISimpleType<EnumOptions>;
};

export const enumeration: EnumerationFactory = <EnumOptions extends string>(
  nameOrOptions: readonly EnumOptions[] | string,
  options?: readonly EnumOptions[],
): ISimpleType<EnumOptions> => {
  let name;
  if (typeof nameOrOptions == "string") {
    name = nameOrOptions;
    options ??= [];
  } else {
    name = "enumeration";
    options = nameOrOptions;
  }
  return new EnumerationType(name, options);
};
