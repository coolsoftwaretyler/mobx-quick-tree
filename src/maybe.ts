import { types as mstTypes } from "mobx-state-tree";
import { BaseType } from "./base";
import { ensureRegistered } from "./class-model";
import type {
  IAnyStateTreeNode,
  IAnyType,
  IMaybeNullType,
  IMaybeType,
  IStateTreeNode,
  InstanceWithoutSTNTypeForType,
  TreeContext,
} from "./types";

export class MaybeType<Type extends IAnyType> extends BaseType<
  Type["InputType"] | undefined,
  Type["OutputType"] | undefined,
  InstanceWithoutSTNTypeForType<Type> | undefined
> {
  constructor(readonly type: Type) {
    super(mstTypes.maybe(type.mstType));
  }

  instantiate(snapshot: this["InputType"] | undefined, context: TreeContext, parent: IStateTreeNode | null): this["InstanceType"] {
    if (snapshot === undefined) {
      return undefined;
    }
    return this.type.instantiate(snapshot, context, parent);
  }

  is(value: any): value is this["InputType"] | this["InstanceType"] {
    if (value === undefined) {
      return true;
    }
    return this.type.is(value);
  }

  async schemaHash() {
    return `maybe:${await this.type.schemaHash()}`;
  }
}

export class MaybeNullType<Type extends IAnyType> extends BaseType<
  Type["InputType"] | null | undefined,
  Type["OutputType"] | null,
  InstanceWithoutSTNTypeForType<Type> | null
> {
  constructor(readonly type: Type) {
    super(mstTypes.maybeNull(type.mstType));
  }

  instantiate(snapshot: this["InputType"] | undefined, context: TreeContext, parent: IStateTreeNode | null): this["InstanceType"] {
    if (snapshot === undefined || snapshot === null) {
      // Special case for things like types.frozen, or types.literal(undefined), where MST prefers the subtype over maybeNull
      if (this.type.is(snapshot)) {
        return this.type.instantiate(snapshot, context, parent);
      }

      return null;
    }
    return this.type.instantiate(snapshot, context, parent);
  }

  is(value: IAnyStateTreeNode): value is this["InstanceType"];
  is(value: any): value is this["InputType"] | this["InstanceType"] {
    if (value === undefined || value === null) {
      return true;
    }
    return this.type.is(value);
  }

  async schemaHash() {
    return `maybeNull:${await this.type.schemaHash()}`;
  }
}

export const maybe = <T extends IAnyType>(type: T): IMaybeType<T> => {
  ensureRegistered(type);
  return new MaybeType(type);
};

export const maybeNull = <T extends IAnyType>(type: T): IMaybeNullType<T> => {
  ensureRegistered(type);
  return new MaybeNullType(type);
};
