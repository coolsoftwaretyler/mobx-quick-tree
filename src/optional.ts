import { types as mstTypes } from "mobx-state-tree";
import { BaseType } from "./base";
import { ensureRegistered } from "./class-model";
import type {
  CreateTypes,
  IAnyStateTreeNode,
  IAnyType,
  InstanceWithoutSTNTypeForType,
  TreeContext,
  IOptionalType,
  IStateTreeNode,
  ValidOptionalValue,
} from "./types";

export type DefaultFuncOrValue<T extends IAnyType> = T["InputType"] | T["OutputType"] | (() => CreateTypes<T>);

export class OptionalType<
  T extends IAnyType,
  OptionalValues extends [ValidOptionalValue, ...ValidOptionalValue[]] = [undefined],
> extends BaseType<T["InputType"] | OptionalValues[number], T["OutputType"], InstanceWithoutSTNTypeForType<T>> {
  constructor(
    readonly type: T,
    readonly defaultValueOrFunc: DefaultFuncOrValue<T>,
    readonly undefinedValues: OptionalValues = [undefined] as any,
  ) {
    super(
      undefinedValues
        ? mstTypes.optional(type.mstType, defaultValueOrFunc, undefinedValues)
        : mstTypes.optional(type.mstType, defaultValueOrFunc),
    );
  }

  instantiate(snapshot: this["InputType"], context: TreeContext, parent: IStateTreeNode | null): this["InstanceType"] {
    if (this.undefinedValues) {
      if (this.undefinedValues.includes(snapshot)) {
        snapshot = this.defaultValue;
      }
    } else if (snapshot === undefined) {
      snapshot = this.defaultValue;
    }

    return this.type.instantiate(snapshot, context, parent);
  }

  is(value: IAnyStateTreeNode): value is this["InstanceType"];
  is(value: any): value is this["InputType"] | this["InstanceType"] {
    if (this.undefinedValues) {
      if (this.undefinedValues.includes(value)) {
        return true;
      }
    } else if (value === undefined) {
      return true;
    }

    return this.type.is(value);
  }

  async schemaHash() {
    return `optional:${await this.type.schemaHash()}`;
  }

  private get defaultValue(): T["InputType"] {
    return this.defaultValueOrFunc instanceof Function ? this.defaultValueOrFunc() : this.defaultValueOrFunc;
  }
}

export type OptionalFactory = {
  <T extends IAnyType>(type: T, defaultValue: DefaultFuncOrValue<T>): IOptionalType<T, [undefined]>;
  <T extends IAnyType, OptionalValues extends [ValidOptionalValue, ...ValidOptionalValue[]]>(
    type: T,
    defaultValue: DefaultFuncOrValue<T>,
    undefinedValues: OptionalValues,
  ): IOptionalType<T, OptionalValues>;
};

export const optional: OptionalFactory = <T extends IAnyType, OptionalValues extends [ValidOptionalValue, ...ValidOptionalValue[]]>(
  type: T,
  defaultValue: DefaultFuncOrValue<T>,
  undefinedValues?: OptionalValues,
): IOptionalType<T, OptionalValues> => {
  ensureRegistered(type);
  return new OptionalType(type, defaultValue, undefinedValues);
};
