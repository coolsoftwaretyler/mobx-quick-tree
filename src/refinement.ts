import type { Instance } from "mobx-state-tree";
import { types } from "mobx-state-tree";
import { BaseType } from "./base";
import type { IAnyType, InstanceWithoutSTNTypeForType, TreeContext, IStateTreeNode, IType } from "./types";

class RefinementType<T extends IAnyType> extends BaseType<T["InputType"], T["OutputType"], InstanceWithoutSTNTypeForType<T>> {
  constructor(
    readonly type: T,
    readonly predicate: (snapshot: Instance<T> | Instance<T["mstType"]>) => boolean,
  ) {
    super(types.refinement(type.mstType, predicate));
  }

  instantiate(snapshot: this["InputType"] | undefined, context: TreeContext, parent: IStateTreeNode | null): this["InstanceType"] {
    const instance = this.type.instantiate(snapshot, context, parent);
    if (!this.predicate(instance)) {
      throw new Error("given snapshot isn't valid for refinement type");
    }
    return instance;
  }

  is(value: any): value is this["InstanceType"];
  is(value: any): value is this["InputType"] | this["InstanceType"] {
    return this.type.is(value) && this.predicate(value);
  }

  async schemaHash() {
    return `refinement:${await this.type.schemaHash()}`;
  }
}

export const refinement = <T extends IAnyType>(
  type: T,
  predicate: (snapshot: Instance<T> | Instance<T["mstType"]>) => boolean,
): IType<T["InputType"], T["OutputType"], InstanceWithoutSTNTypeForType<T>> => {
  return new RefinementType(type, predicate);
};
