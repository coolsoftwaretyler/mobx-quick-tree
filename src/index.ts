import { types as mstTypes } from "mobx-state-tree";
import { array } from "./array";
import { compose } from "./compose";
import { custom } from "./custom";
import { enumeration } from "./enumeration";
import { frozen } from "./frozen";
import { late } from "./late";
import { map } from "./map";
import { maybe, maybeNull } from "./maybe";
import { model } from "./model";
import { optional } from "./optional";
import { reference, safeReference } from "./reference";
import { refinement } from "./refinement";
import { DateType, IntegerType, NullType, SimpleType, literal } from "./simple";
import { lazyUnion, union } from "./union";

export * from "./api";
export * from "./types";
export { $type } from "./symbols";
export { setDefaultShouldEmitPatchOnChange } from "./class-model";

export const types = {
  boolean: SimpleType.for("boolean", mstTypes.boolean),
  Date: new DateType(mstTypes.Date),
  identifier: SimpleType.for("string", mstTypes.identifier),
  integer: new IntegerType(),
  literal,
  null: new NullType(),
  number: SimpleType.for("number", mstTypes.number),
  string: SimpleType.for("string", mstTypes.string),

  array,
  compose,
  custom,
  enumeration,
  frozen,
  late,
  lazyUnion,
  map,
  maybe,
  maybeNull,
  model,
  optional,
  reference,
  refinement,
  safeReference,
  union,
};
