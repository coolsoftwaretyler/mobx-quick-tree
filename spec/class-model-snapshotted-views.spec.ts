import { observable, runInAction } from "mobx";
import { ClassModel, action, snapshottedView, getSnapshot, register, types, onPatch } from "../src";
import { Apple } from "./fixtures/FruitAisle";
import { create } from "./helpers";
import { getParent } from "mobx-state-tree";
import { setDefaultShouldEmitPatchOnChange } from "../src/class-model";

@register
class ViewExample extends ClassModel({ key: types.identifier, name: types.string }) {
  @snapshottedView()
  get slug() {
    return this.name.toLowerCase().replace(/ /g, "-");
  }

  @action
  setName(name: string) {
    this.name = name;
  }
}

@register
class Outer extends ClassModel({ name: types.string, examples: types.array(ViewExample) }) {
  @snapshottedView()
  get upperName() {
    return this.name.toUpperCase();
  }
}

describe("class model snapshotted views", () => {
  describe.each([
    ["read-only", true],
    ["observable", false],
  ])("%s", (_name, readOnly) => {
    test("instances don't require the snapshot to include the cache", () => {
      const instance = create(ViewExample, { key: "1", name: "Test" }, readOnly);
      expect(instance.slug).toEqual("test");
    });

    test("models with cached views still correctly report .is on totally different models", () => {
      const instance = create(ViewExample, { key: "1", name: "Test" }, readOnly);
      expect(ViewExample.is(instance)).toBe(true);
      expect(Apple.is(instance)).toBe(false);

      const other = create(Apple, { type: "Apple", ripeness: 1 }, readOnly);
      expect(ViewExample.is(other)).toBe(false);
      expect(Apple.is(other)).toBe(true);
    });

    test("instances of models with all optional properties arent .is of other models with all optional properties", () => {
      @register
      class AllOptionalA extends ClassModel({ name: types.optional(types.string, "Jim") }) {}

      @register
      class AllOptionalB extends ClassModel({ title: types.optional(types.string, "Jest") }) {}

      // the empty snapshot matches both types
      expect(AllOptionalA.is({})).toBe(true);
      expect(AllOptionalB.is({})).toBe(true);

      const instanceA = create(AllOptionalA, {}, readOnly);
      expect(AllOptionalA.is(instanceA)).toBe(true);
      expect(AllOptionalB.is(instanceA)).toBe(false);

      const instanceB = create(AllOptionalA, {}, readOnly);
      expect(AllOptionalA.is(instanceB)).toBe(true);
      expect(AllOptionalB.is(instanceB)).toBe(false);
    });
  });

  test("an observable instance ignores the input snapshot value as the logic may have changed", () => {
    const instance = ViewExample.create({ key: "1", name: "Test", slug: "outdated-cache" } as any);
    expect(instance.slug).toEqual("test");
  });

  test("an observable instance's snapshot includes the snapshotted views epoch", () => {
    const instance = ViewExample.create({ key: "1", name: "Test" });
    expect(getSnapshot(instance)).toEqual({ __snapshottedViewsEpoch: 0, key: "1", name: "Test" });
  });

  test("a readonly instance's snapshot doesn't include the snapshotted views epoch", () => {
    const instance = ViewExample.createReadOnly({ key: "1", name: "Test" });
    expect(getSnapshot(instance)).toEqual({ key: "1", name: "Test" });
  });

  test("a readonly instance returns the view value from the snapshot if present", () => {
    const instance = ViewExample.createReadOnly({ key: "1", name: "Test", slug: "test" } as any);
    expect(instance.slug).toEqual("test");
  });

  test("a readonly instance doesn't recompute the view value from the snapshot", () => {
    const instance = ViewExample.createReadOnly({ key: "1", name: "Test", slug: "whatever" } as any);
    expect(instance.slug).toEqual("whatever");
  });

  test("a readonly instance doesn't call the computed function if given a snapshot value", () => {
    const fn = jest.fn();
    @register
    class Spy extends ClassModel({ name: types.string }) {
      @snapshottedView()
      get slug() {
        fn();
        return this.name.toLowerCase().replace(/ /g, "-");
      }
    }

    const instance = Spy.createReadOnly({ name: "Test", slug: "whatever" } as any);
    expect(instance.slug).toEqual("whatever");
    expect(fn).not.toHaveBeenCalled();
  });

  test("snapshotted views can be passed nested within snapshots", () => {
    const instance = Outer.createReadOnly({
      name: "foo",
      upperName: "SNAPSHOT",
      examples: [{ key: "1", name: "Test", slug: "test-foobar" } as any, { key: "2", name: "Test 2", slug: "test-qux" } as any],
    } as any);

    expect(instance.upperName).toEqual("SNAPSHOT");
    expect(instance.examples[0].slug).toEqual("test-foobar");
    expect(instance.examples[1].slug).toEqual("test-qux");
  });

  describe("with a hydrator", () => {
    @register
    class HydrateExample extends ClassModel({ url: types.string }) {
      @snapshottedView<URL>({
        createReadOnly(value, node) {
          expect(node).toBeDefined();
          return value ? new URL(value) : undefined;
        },
      })
      get withoutParams() {
        const url = new URL(this.url);
        for (const [key] of url.searchParams.entries()) {
          url.searchParams.delete(key);
        }
        return url;
      }

      @action
      setURL(url: string) {
        this.url = url;
      }
    }

    test("snapshotted views with processors can be accessed on observable instances", () => {
      const instance = HydrateExample.create({ url: "https://gadget.dev/blog/feature?utm=whatever" });
      expect(instance.withoutParams).toEqual(new URL("https://gadget.dev/blog/feature"));
    });

    test("snapshotted views with processors can be accessed on readonly instances when there's no input data", () => {
      const instance = HydrateExample.create({ url: "https://gadget.dev/blog/feature?utm=whatever" });
      expect(instance.withoutParams).toEqual(new URL("https://gadget.dev/blog/feature"));
    });

    test("snapshotted views with processors can be accessed on readonly instances when there is input data", () => {
      const instance = HydrateExample.createReadOnly({
        url: "https://gadget.dev/blog/feature?utm=whatever",
        withoutParams: "https://gadget.dev/blog/feature/extra", // pass a different value so we can be sure it is what is being used
      } as any);
      expect(instance.withoutParams).toEqual(new URL("https://gadget.dev/blog/feature/extra"));
    });

    test("hydrators aren't called eagerly on readonly instances in case they are expensive", () => {
      const fn = jest.fn().mockReturnValue("whatever");
      @register
      class HydrateExampleSpy extends ClassModel({}) {
        @snapshottedView<URL>({
          createReadOnly: fn,
        })
        get someView() {
          return "view value";
        }
      }

      const instance = HydrateExampleSpy.createReadOnly({ someView: "snapshot value" });
      expect(fn).not.toHaveBeenCalled();

      expect(instance.someView).toEqual("whatever");
      expect(fn).toHaveBeenCalledTimes(1);
    });
  });

  describe("references", () => {
    @register
    class Referencer extends ClassModel({
      id: types.identifier,
      example: types.reference(ViewExample),
    }) {}

    @register
    class Root extends ClassModel({
      referrers: types.map(Referencer),
      examples: types.map(ViewExample),
    }) {}

    test("references to models with snapshotted views can be instantiated", () => {
      const root = Root.createReadOnly({
        referrers: {
          a: { id: "a", example: "1" },
          b: { id: "b", example: "2" },
        },
        examples: {
          "1": { key: "1", name: "Alice" },
          "2": { key: "2", name: "Bob" },
        },
      });

      expect(getSnapshot(root)).toMatchSnapshot();
    });
  });

  describe("parents", () => {
    const onError = jest.fn();

    @register
    class Child extends ClassModel({}) {
      @snapshottedView({ onError, shouldEmitPatchOnChange: () => true })
      get parentsChildLength() {
        const parent: Parent = getParent(this, 2);
        return parent.children.size;
      }
    }

    @register
    class Parent extends ClassModel({ children: types.map(Child) }) {
      @action
      setChild(key: string, child: Child) {
        this.children.set(key, child);
      }

      @action
      createChild(key: string) {
        this.children.set(key, {});
        return this.children.get(key);
      }

      @action
      removeChild(key: string) {
        this.children.delete(key);
      }
    }

    test("observable instances with views that access their parent will cause their view's reaction to error when created outside an action", () => {
      Child.create();
      expect(onError).toHaveBeenCalled();
    });

    test("observable instances with views that access their parent will not cause their view's reaction to error when created inside an action", () => {
      runInAction(() => {
        const child = Child.create();
        const parent = Parent.create();
        parent.setChild("foo", child);
      });

      expect(onError).not.toHaveBeenCalled();
    });

    test("observable instances don't run their view's reaction when they are destroyed", () => {
      const parent = Parent.create();
      parent.createChild("foo");
      parent.removeChild("foo");

      // would have errored if child.parentsChildLength was accessed after the child was removed
      expect(onError).not.toHaveBeenCalled();
    });
  });

  describe("shouldEmitPatchOnChange", () => {
    afterEach(() => {
      // reset the default value
      setDefaultShouldEmitPatchOnChange(false);
    });

    test("readonly instances don't use the shouldEmitPatchOnChange option", () => {
      const fn = jest.fn();
      @register
      class MyViewExample extends ClassModel({ key: types.identifier, name: types.string }) {
        @snapshottedView({ shouldEmitPatchOnChange: fn })
        get slug() {
          return this.name.toLowerCase().replace(/ /g, "-");
        }
      }

      const instance = MyViewExample.createReadOnly({ key: "1", name: "Test" });
      expect(instance.slug).toEqual("test");
      expect(fn).not.toHaveBeenCalled();
    });

    test("observable instances don't emit a patch when shouldEmitPatchOnChange returns false", () => {
      const shouldEmitPatchOnChangeFn = jest.fn(() => false);
      const observableArray = observable.array<string>([]);

      @register
      class MyViewExample extends ClassModel({ key: types.identifier, name: types.string }) {
        @snapshottedView({ shouldEmitPatchOnChange: shouldEmitPatchOnChangeFn })
        get arrayLength() {
          return observableArray.length;
        }
      }

      const instance = MyViewExample.create({ key: "1", name: "Test" });
      expect(shouldEmitPatchOnChangeFn).toHaveBeenCalled();

      const onPatchFn = jest.fn();
      onPatch(instance, onPatchFn);

      runInAction(() => {
        observableArray.push("a");
      });

      expect(onPatchFn).not.toHaveBeenCalled();
    });

    test("observable instances do emit a patch when shouldEmitPatchOnChange returns true", () => {
      const shouldEmitPatchOnChangeFn = jest.fn(() => true);
      const observableArray = observable.array<string>([]);

      @register
      class MyViewExample extends ClassModel({ key: types.identifier, name: types.string }) {
        @snapshottedView({ shouldEmitPatchOnChange: shouldEmitPatchOnChangeFn })
        get arrayLength() {
          return observableArray.length;
        }
      }

      const instance = MyViewExample.create({ key: "1", name: "Test" });
      expect(shouldEmitPatchOnChangeFn).toHaveBeenCalled();

      const onPatchFn = jest.fn();
      onPatch(instance, onPatchFn);

      runInAction(() => {
        observableArray.push("a");
      });

      expect(onPatchFn).toHaveBeenCalled();
      expect(onPatchFn.mock.calls).toEqual([
        [
          { op: "replace", path: "/__snapshottedViewsEpoch", value: 1 },
          { op: "replace", path: "/__snapshottedViewsEpoch", value: 0 },
        ],
      ]);
    });

    test("observable instances don't emit a patch when shouldEmitPatchOnChange is undefined and setDefaultShouldEmitPatchOnChange hasn't been called", () => {
      const observableArray = observable.array<string>([]);

      @register
      class MyViewExample extends ClassModel({ key: types.identifier, name: types.string }) {
        @snapshottedView()
        get arrayLength() {
          return observableArray.length;
        }
      }

      const instance = MyViewExample.create({ key: "1", name: "Test" });

      const onPatchFn = jest.fn();
      onPatch(instance, onPatchFn);

      runInAction(() => {
        observableArray.push("a");
      });

      expect(onPatchFn).not.toHaveBeenCalled();
    });

    test("observable instances do emit a patch when shouldEmitPatchOnChange is undefined and setDefaultShouldEmitPatchOnChange was passed true", () => {
      setDefaultShouldEmitPatchOnChange(true);

      const observableArray = observable.array<string>([]);

      @register
      class MyViewExample extends ClassModel({ key: types.identifier, name: types.string }) {
        @snapshottedView()
        get arrayLength() {
          return observableArray.length;
        }
      }

      const instance = MyViewExample.create({ key: "1", name: "Test" });

      const onPatchFn = jest.fn();
      onPatch(instance, onPatchFn);

      runInAction(() => {
        observableArray.push("a");
      });

      expect(onPatchFn).toHaveBeenCalled();
      expect(onPatchFn).toHaveBeenCalled();
      expect(onPatchFn.mock.calls).toEqual([
        [
          { op: "replace", path: "/__snapshottedViewsEpoch", value: 1 },
          { op: "replace", path: "/__snapshottedViewsEpoch", value: 0 },
        ],
      ]);
    });

    test("observable instances don't emit a patch when shouldEmitPatchOnChange is undefined and setDefaultShouldEmitPatchOnChange was passed false", () => {
      setDefaultShouldEmitPatchOnChange(false);

      const observableArray = observable.array<string>([]);

      @register
      class MyViewExample extends ClassModel({ key: types.identifier, name: types.string }) {
        @snapshottedView()
        get arrayLength() {
          return observableArray.length;
        }
      }

      const instance = MyViewExample.create({ key: "1", name: "Test" });

      const onPatchFn = jest.fn();
      onPatch(instance, onPatchFn);

      runInAction(() => {
        observableArray.push("a");
      });

      expect(onPatchFn).not.toHaveBeenCalled();
    });
  });
});
