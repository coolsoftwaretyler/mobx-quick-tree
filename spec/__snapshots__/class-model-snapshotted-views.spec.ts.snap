// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class model snapshotted views an observable instance emits a patch when the view value changes 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "op": "replace",
        "path": "/__snapshottedViewsEpoch",
        "value": 1,
      },
      {
        "op": "replace",
        "path": "/__snapshottedViewsEpoch",
        "value": 0,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`class model snapshotted views references references to models with snapshotted views can be instantiated 1`] = `
{
  "examples": {
    "1": {
      "key": "1",
      "name": "Alice",
    },
    "2": {
      "key": "2",
      "name": "Bob",
    },
  },
  "referrers": {
    "a": {
      "example": "1",
      "id": "a",
    },
    "b": {
      "example": "2",
      "id": "b",
    },
  },
}
`;
